cmd/helm/downloader/chart_downloader.go

```diff
@@ -132,63 +131,75 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
 //		* If version is non-empty, this will return the URL for that version
 //		* If version is empty, this will return the URL for the latest version
 // 		* If no version can be found, an error is returned
-func (c *ChartDownloader) ResolveChartVersion(ref, version string) (*url.URL, error) {
-	// See if it's already a full URL.
-	// FIXME: Why do we use url.ParseRequestURI instead of url.Parse?
-	u, err := url.ParseRequestURI(ref)
-	if err == nil {
-		// If it has a scheme and host and path, it's a full URL
-		if u.IsAbs() && len(u.Host) > 0 && len(u.Path) > 0 {
-			return u, nil
-		}
-		return u, fmt.Errorf("invalid chart url format: %s", ref)
+func (c *ChartDownloader) ResolveChartVersion(ref, version string) (*url.URL, *http.Client, error) {
+	u, err := url.Parse(ref)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid chart URL format: %s", ref)
 	}
 
-	r, err := repo.LoadRepositoriesFile(c.HelmHome.RepositoryFile())
+	rf, err := repo.LoadRepositoryFile(c.HelmHome.RepositoryFile())
 	if err != nil {
-		return u, err
+		return nil, nil, err
 	}
 
-	// See if it's of the form: repo/path_to_chart
-	p := strings.SplitN(ref, "/", 2)
-	if len(p) < 2 {
-		return u, fmt.Errorf("invalid chart url format: %s", ref)
+	var (
+		chartName string
+		rc        *repo.ChartRepositoryConfig
+	)
+	if u.IsAbs() && len(u.Host) > 0 && len(u.Path) > 0 {
+		// If it has a scheme and host and path, it's a full URL
+		p := strings.SplitN(strings.TrimLeft(u.Path, "/"), "-", 2)
+		if len(p) < 2 {
+			return nil, nil, fmt.Errorf("Seems that chart path is not in form of repo_url/path_to_chart, got: %s", u)
+		}
+		chartName = p[0]
+		u.Path = ""
+		rc, err = pickChartRepositoryConfigByURL(u.String(), rf.Repositories)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		// See if it's of the form: repo/path_to_chart
+		p := strings.SplitN(u.Path, "/", 2)
+		if len(p) < 2 {
+			return nil, nil, fmt.Errorf("Non-absolute URLs should be in form of repo_name/path_to_chart, got: %s", u)
+		}
+
+		repoName := p[0]
+		chartName = p[1]
+		rc, err = pickChartRepositoryConfigByName(repoName, rf.Repositories)
+		if err != nil {
+			return nil, nil, err
+		}
 	}
 
-	repoName := p[0]
-	chartName := p[1]
-	rf, err := findRepoEntry(repoName, r.Repositories)
+	r, err := repo.NewChartRepository(rc)
```

Comment:
This is just a suggestion that Adam and I think would make this a little more straightforward.

How about instead of having `NewChartRepository` create the client, we add a method to the ChartRepository that would be something like `Get(url) ([]byte, error)`? So instead of returning the client, the `ChartRepository` object would simply abstract away that logic.

Since `http.Get` and `ChartRepository.Get` would then have the same method signature, we could give `download` a function signature like `download(url string, client Getter)`.